GC problems in 4.11.0

I am trying to build GAP 4.11.0 for Fedora Rawhide.  I am having issues with the garbage collector.  The initial build used the default gasman collector.  That worked fine on all architectures except aarch64 (the 64-bit ARM architecture).  GAP 4.10.2 did not have any issues on aarch64.  With 4.11.0, the executable and documentation built successfully, but many tests failed like this:
```
$ make check
/builddir/build/BUILD/gap-4.11.0/bin/gap.sh --quitonbreak -b -m 100m -o 1g -q -x 80 -r -l /builddir/build/BUILD/gap-4.11.0 -A  tst/testinstall.g
You should start GAP4 using `gap -A -x 80 -r -m 100m -o 1g -K 2g'.
Architecture: aarch64-redhat-linux-gnu-default64-kv7
testing: /builddir/build/BUILD/gap-4.11.0/tst/testinstall/ConjNatSym.tst
     148 ms (28 ms GC) and 23.8MB allocated for ConjNatSym.tst
testing: /builddir/build/BUILD/gap-4.11.0/tst/testinstall/DirectProductElement\
.tst
########> Diff in /builddir/build/BUILD/gap-4.11.0/tst/testinstall/DirectProdu\
ctElement.tst:2
# Input is:
START_TEST("DirectProductElement.tst");
# Expected output:
# But found:
Error, Length: <list> must be a list (not a function body bag)
########
     106 ms (91 ms GC) and 655KB allocated for DirectProductElement.tst
testing: /builddir/build/BUILD/gap-4.11.0/tst/testinstall/ElmsBlist.tst
########> Diff in /builddir/build/BUILD/gap-4.11.0/tst/testinstall/ElmsBlist.t\
st:1
# Input is:
START_TEST("ElmsBlist.tst");
# Expected output:
# But found:
Error, Length: <list> must be a list (not a integer)
########
     127 ms (82 ms GC) and 4.33MB allocated for ElmsBlist.tst
```

All of the failures refer to START_TEST.  A little experimentation shows that the problem happens at this line in START_TEST: `GASMAN( "collect" );`.  I can try this week to get access to aarch64 hardware to debug the problem, but would appreciate any ideas on what might be going wrong.

I considered using an alternate GC.  Julia is only available on x86, so that doesn't help.  I built with Boehm GC.  That failed because we are now using gcc 10, which defaults to -fno-common, and the version of Boehm GC bundled by GAP has multiple definitions of the same variable.  I added -fcommon to the build flags, and the build succeeded on all 64-bit architectures, including aarch64, but failed on both 32-bit architectures (i686 and 32-bit ARM).  On both platforms, the documentation building step failed:
```
$ make manuals
doc/make_doc
--------------------
Building GAP manuals
--------------------
Error before error-handling is initialized: 
[ 
  "Record: '<rec>.(<obj>)' <obj> must be a string or a small integer (not a lar\
make: *** [Makefile.rules:791: doc] Error 1
```

So my last resort, if none of these problems can be debugged, is to build with the gasman collector on some architectures (i686 and 32-bit ARM, at least) and boehm gc on others (aarch64, at least).  Any guidance or suggestions you can offer are much appreciated.  Thank you.

I haven't yet managed to get direct access to an aarch64 machine, I'm afraid.  Fedora has a couple of test machines that are supposed to be used for such purposes, but both have been rejecting ssh connection attempts for the last few days.  I have filed an issue with the Fedora staff asking for this to be fixed, but haven't heard back yet.  If I don't hear from them in the next 12 hours or so, I will try to do a build with qemu myself.  Right now, all I can do is lob sources at the Fedora build machines and then read the build logs afterward.

I suspect that this behavior will only be reproducible with gcc 10, which Fedora recently switched to.  It may even prove to be an aarch64-specific bug in gcc 10, but that will require some investigation.

I did a few test builds today, and can now tell you that adding --enable-debug makes the problem go away.  I suspected that this was due to disabling __attribute__((always_inline)), so I tried patching system.h to do that directly.  However, the build failed the same way.  I also tried reducing the optimization level to -O1, but that made the documentation building step fail like this:
```
#I Finally the HTML version . . .
#I First run, collecting cross references, index, toc, bib and so on . . .
#I Table of contents complete.
#I Producing the index . . .
#I Writing bibliography . . .
#I Second run through document . . .
Panic in src/gasman.c:2120: Gasman found a bogus header
make: *** [Makefile.rules:791: doc] Error 1
```
which again makes me suspect a gcc bug.  I'll need a machine where I can run a debugger to really figure out what is going on though.  Hopefully tomorrow....

--- a/src/gasman.c
+++ b/src/gasman.c
@@ -1858,8 +1858,7 @@ static void SparcStackFuncBags(void)
 }
 #endif
 
-
-static void GenStackFuncBags(void)
+static NOINLINE void GenStackFuncBags(void)
 {
     Bag *               top;            /* top of stack                    */
     Bag *               p;              /* loop variable                   */

--- a/src/hpc/thread.c
+++ b/src/hpc/thread.c
@@ -254,12 +254,6 @@ static void RemoveGCRoots(void)
 
 #if !defined(USE_NATIVE_TLS) && !defined(USE_PTHREAD_TLS)
 
-#ifdef HAVE_FUNC_ATTRIBUTE_NOINLINE
-#define NOINLINE __attribute__((noinline))
-#else
-#define NOINLINE
-#endif
-
 
 /* In order to safely use thread-local memory on the main stack, we have
  * to work around an idiosyncracy in some virtual memory systems. These

--- a/src/system.h
+++ b/src/system.h
@@ -63,6 +63,31 @@ enum {
 *T  Wrappers for various compiler attributes
 **
 */
+
+// recent clang and gcc versions have __has_attribute; for compilers that lack
+// it, we have to rely on the autoconf test results.
+#ifdef __has_attribute
+
+#if __has_attribute(always_inline)
+#define HAVE_FUNC_ATTRIBUTE_ALWAYS_INLINE 1
+#else
+#undef HAVE_FUNC_ATTRIBUTE_ALWAYS_INLINE
+#endif
+
+#if __has_attribute(noreturn)
+#define HAVE_FUNC_ATTRIBUTE_NORETURN 1
+#else
+#undef HAVE_FUNC_ATTRIBUTE_NORETURN
+#endif
+
+#if __has_attribute(noinline)
+#define HAVE_FUNC_ATTRIBUTE_NOINLINE 1
+#else
+#undef HAVE_FUNC_ATTRIBUTE_NOINLINE
+#endif
+
+#endif
+
 #if defined(HAVE_FUNC_ATTRIBUTE_ALWAYS_INLINE) && !defined(GAP_KERNEL_DEBUG)
 #define ALWAYS_INLINE __attribute__((always_inline)) inline
 #else
@@ -73,6 +98,12 @@ enum {
 #define NORETURN __attribute__((noreturn))
 #else
 #define NORETURN
+#endif
+
+#ifdef HAVE_FUNC_ATTRIBUTE_NOINLINE
+#define NOINLINE __attribute__((noinline))
+#else
+#define NOINLINE
 #endif
 
 

